### DISCUSSION TITLE
Unpackaged publishing

### DISCUSSION THREAD

**Discussion by Petrarca181** [microsoft/microsoft-ui-xaml#8389] [Q&A]
Hello, when I  compile/publish my app, visual studio creates a lot of language folders like "en-US"  for every language.  Are they actually needed if my app is  english only?  And how  can I stop visual studio create them?

At this moment any possibility to publish single file for unpackaged self contained + appsdk self contained?  I'm  tired to see 350 dlls in my app folder :)

**Comment by DarranRowe** [comment]
For the languages, when your application is self contained, what actually happens is that Visual Studio copies the contents of the Windows App SDK framework package to your application directory. This is the file named Microsoft.WindowsAppRuntime.1.x.msix. So you can't really stop the creation of the language directories. For whether or not you can delete them, technically you can. They contain .mui resources, the same as Windows' language packs. But this would essentially be unsupported.

As for the ability to publish a single file, the answer will most likely always be no. The Windows App SDK components are native meaning that these will be separate files.

**Reply by Petrarca181** [reply]
> For the languages, when your application is self contained, what actually happens is that Visual Studio copies the contents of the Windows App SDK framework package to your application directory. This is the file named Microsoft.WindowsAppRuntime.1.x.msix. So you can't really stop the creation of the language directories. For whether or not you can delete them, technically you can. They contain .mui resources, the same as Windows' language packs. But this would essentially be unsupported.
> 
> As for the ability to publish a single file, the answer will most likely always be no. The Windows App SDK components are native meaning that these will be separate files.

Thank you. Technically possible to move framework and sdk dlls  into different folder or they must be in the same folder with .exe? Like it's possible with net.core https://github.com/nulastudio/NetBeauty2

**Reply by DarranRowe** [reply]
A bit more of an explanation is needed for the Windows App SDK libraries.
These are Windows Runtime (WinRT) components. For Windows Runtime components, there has to be some form of registration of these components to be usable. When it is in a package, the registration occurs through the package itself. However, when there is no package, another means of registration is needed.
This is where a Windows feature called "Registration Free WinRT" comes in. This puts information into the application manifest file that gives information about what runtime classes are available and what library it is in. This is where the difficulty is, as you can see from [this blog post introducing this feature](https://blogs.windows.com/windowsdeveloper/2019/04/30/enhancing-non-packaged-desktop-apps-using-windows-runtime-components/), it uses the file element of the application manifest file, and it only gives a file name. What's more, [the documentation for the file element in the manifest file](https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#file) states that it only takes a file name. For verification, I extracted the embedded manifest from an application set to self contained and it did contain the entries for the WinRT components.
So on its own, it is not possible. But again, there are two other Windows features that could work here, but I honestly haven't tested them out. First is the older [DLL redirection](https://learn.microsoft.com/en-us/windows/win32/sbscs/dll-com-redirection-on-windows). This is disabled by default when an application manifest is present, but you should be able to activate it again by modifying the registry. The second is [private side by side assemblies](https://learn.microsoft.com/en-us/windows/win32/sbscs/dll-com-redirection-on-windows). But as I wrote, I haven't tested these, so I'm not sure if they will work. But you should assume no.
I'm not sure about the .NET framework, maybe someone who is familiar with the .NET framework can provide more information on that.

**Reply by Petrarca181** [reply]
I think it will be faster to put the main winui 3  app into some folder like "bin" or "core"  and create a dummy sub app to run the main one  and manage files. This will make things cleaner.

**Comment by GuildOfCalamity** [comment]
I've modified this example from an **AndrewKeepCoding** repo (edit your _csproj_ file)...
```
<Project>
  <!-- This target will remove the language folders inside the build output except the specified one. -->
  <Target Name="RemoveFoldersWithMuiFiles" Condition="'$(WindowsPackageType)' == 'None'" AfterTargets="Build">
        <ItemGroup>
            <RemovingFiles Include="$(OutDir)*\*.mui" Exclude="$(OutDir)en-us\*.mui" />
            <RemovingFolders Include="@(RemovingFiles->'%(RootDir)%(Directory)')" />
        </ItemGroup>
        <RemoveDir Directories="@(RemovingFolders)" />
  </Target>
</Project>
```
You can edit this to match both **Packaged** and **Unpackaged** builds.
_NOTE:_ If the language subfolder contains no `*.mui` contents then it will still be there afterwards.
If you want to clean those also then you can modify the script like so...
```
  <Target Name="RemoveFoldersWithMuiFiles" Condition="'$(WindowsPackageType)' == 'None'" AfterTargets="Build">
        <ItemGroup>
            <RemovingFiles Include="$(OutDir)*\*.mui" Exclude="$(OutDir)en-us\*.mui" />
            <RemovingFolders Include="@(RemovingFiles->'%(RootDir)%(Directory)')" />
            <RemovingCS Include="$(OutDir)cs" />
            <RemovingDE Include="$(OutDir)de" />
            <RemovingES Include="$(OutDir)es" />
            <RemovingFR Include="$(OutDir)fr" />
            <RemovingIT Include="$(OutDir)it" />
            <RemovingJA Include="$(OutDir)ja" />
            <RemovingKO Include="$(OutDir)ko" />
            <RemovingPL Include="$(OutDir)pl" />
            <RemovingRU Include="$(OutDir)ru" />
            <RemovingTR Include="$(OutDir)tr" />
            <RemovingZHS Include="$(OutDir)zh-Hans" />
            <RemovingZHT Include="$(OutDir)zh-Hant" />
        </ItemGroup>
        <RemoveDir Directories="@(RemovingFolders)" />
        <RemoveDir Directories="@(RemovingCS)" />
        <RemoveDir Directories="@(RemovingDE)" />
        <RemoveDir Directories="@(RemovingES)" />
        <RemoveDir Directories="@(RemovingFR)" />
        <RemoveDir Directories="@(RemovingIT)" />
        <RemoveDir Directories="@(RemovingJA)" />
        <RemoveDir Directories="@(RemovingKO)" />
        <RemoveDir Directories="@(RemovingPL)" />
        <RemoveDir Directories="@(RemovingRU)" />
        <RemoveDir Directories="@(RemovingTR)" />
        <RemoveDir Directories="@(RemovingZHS)" />
        <RemoveDir Directories="@(RemovingZHT)" />
  </Target>
```
